packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+examples can be found at the relevant client docs).
+
+An alternative to this way of doing things is to overload a command which all WebDriver clients
+have access to already: Execute Script. A common driver pattern is to create a magic prefix which
+the driver looks for at the beginning of any Execute Script invocation, and if it's present, to
+treat the rest of the script string as the name of a custom command to execute. This works in large
+part because for most platforms beyond web browsers, executing arbitrary JavaScript isn't a thing
+that makes sense.
+
+Let's say you are building a driver for stereo system called `soundz`, and you wanted to create
+a command for playing a song by name. You could expose this to your users in such a way that they
+call something like:
+
+```js
+// webdriverio example
+driver.executeScript('soundz: playSong', [{song: 'Stairway to Heaven', artist: 'Led Zeppelin'}]);
+```
+
+Then in your driver code you could simply look for `soundz:` at the beginning of your
+`execute` override, and do the appropriate thing:
+
+```js
+async execute(script, args) {
+  if (script === 'soundz: playSong') {
+    const {song, artist} = args[0];
+    // play the song based on song and artist details
+    return;
+  }
+  // otherwise do the normal thing, or throw a not yet implemented
+}
+```
+
+### Implement handling of Appium settings
+
+Appium users can send parameters to your driver via CLI args as well as via capabilities. But these
+cannot change during the course of a test, and sometimes users want to adjust parameters mid-test.
+Appium has a [Settings](../guides/settings.md) API for this purpose.
+
+To support settings in your own driver, first of all define `this.settings` to be an instance of
+the appropriate class, in your constructor:
+
+```js
+// import {DeviceSettings} from 'appium/driver';
+
+this.settings = new DeviceSettings();
+```
+
+Now, you can read user settings any time simply by calling `this.settings.getSettings()`. This will
+return a JS object where the settings names are keys and have their corresponding values.
+
+If you want to assign some default settings, or run some code on your end whenever settings are
+updated, you can do both of these things as well.
+
+```js
+constructor() {
+  const defaults = {setting1: 'value1'};
+  this.settings = new DeviceSettings(defaults, this.onSettingsUpdate.bind(this));
+}
+
+async onSettingsUpdate(key, value) {
+  // do anything you want here with key and value
+}
+```
+
+### Make itself aware of resources other concurrent drivers are using
+
+Let's say your driver uses up some system resources, like ports. There are a few ways to make sure
+that multiple simultaneous sessions don't use the same resources:
+
+1. Have your users specify resource IDs via capabilities (`appium:driverPort` etc)
+1. Just always use free resources (find a new random port for each session)
+1. Have each driver express what resources it is using, then examine currently-used resources from
```

Comment:
this strategy is only going to work if drivers are triggered in scope of a single server. If a common system resource is being used by drivers running on the same machine but under different server processes then armageddon is guaranteed. In such case the usage of inter-process locks (for example file-based ones) could only be helpful.